Federation Averaging with Differential Privacy

BGluth opened the pull request:
So this is an initial impl. It's definitely not fully correct (but it runs), and there are still some major TODOs. I've added an example of running this in `Using the Runner.ipynb`.

Noteable TODOs:

- Theta is not being calculated correctly in `user_update_fed_avg`.
- _get_random_selection_of_user_idxs is always picking a 1/5th sample of users each time (it probably should be variable).
- Moments accountant stuff is still all a placeholder.
- Not heavily tested!!

BGluth committed:
Initial function breakdown for fed_avg_w_dp impl

BGluth committed:
WIP (SQUASH)

BGluth committed:
Partial initial impl of Fed. Avg w/ DP (Brendan)

- Not runnable yet!
- The rough structure is done though.

BGluth committed:
Impled _gen_gausian_rand_noise

zhaoyuxuan committed:
update federate averaging function

zhaoyuxuan committed:
update server update federated averaging function

BGluth committed:
All stub functions now have some impl

- Some of them just return placeholder values.

BGluth committed:
fed_avg_w_dp now can be run with the runner

- Added the glue (param keys, run function) to the runner for this sim.
- Also added an example for this sim in the notebook.

BGluth committed:
More work on finishing impl for fed_avg_w_dp...

BGluth committed:
fed_avg_w_dp now runs (with missing logic)

- Also updated the notebook.

- Noteable TODOs:
    - Theta is not being calculated correctly in `user_update_fed_avg`.
    - _get_random_selection_of_user_idxs is always picking a 1/5th
    sample of users each time (it probably should be variable).
    - Moments accountant stuff is still all a placeholder.
    - Not really tested!!

BGluth committed:
Added a few comments

BGluth committed:
Fixed theata clipping incorrectness (hopefully)

BGluth committed:
fed_avg_w_dp now runs

- Likely incorrect...

crankycoder commented:
@maharshmellow @zhaoyuxuan  let's get some testcases put into this code so that everyone can review the codebase for understanding and to isolate some early bugs.

BGluth committed:
Desperate attempt to calculate a score for a run

- Issue still exists with slices of coef not being np.array no matter
how hard I try. (See scattering of calls everywhere trying to convert a
python list to a np array)

BGluth committed:
fed_avg_w_dp now outputs a (likely valid) score

BGluth committed:
Moved fed_avg_w_dp into mozfldp

BGluth committed:
Fixed incorrect import in runner after moving fed_avg_w_dp_clipping

BGluth committed:
Fixed all randomness within fed_avg_w_dp

- Note: Still waiting on PR #27 to be merged in so the runner random seed param key exists.

BGluth committed:
fed_avg_w_dp appears to now converge consistantly

BGluth committed:
Python Black pass

BGluth commented:
@crankycoder Ok I think this is ready for review as it now converges on the blob dataset.

BGluth committed:
Removed unused parameter

BGluth committed:
Renamed merge function to indicate that we are merging thetas

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+        self.num_rounds = num_rounds
+        self.batch_size = batch_size
+        self.num_epochs = num_epochs
+        self.weight_mod = weight_mod  # w_hat
```
I would rename this parameter to something more reflective of its purpose, such as `user_weight_cap`

BGluth replied:
Fixed.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+        prms.num_users, prms.weight_mod
+    )
+    theta = None
+    theta_0 = _init_theta_from_moment_accountant(prms.num_features, prms.num_labels)
```
The initialization of the model weights is not related to the moments accountant. I would rename this function to drop the reference.

BGluth replied:
Changed.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    )
+    theta = None
+    theta_0 = _init_theta_from_moment_accountant(prms.num_features, prms.num_labels)
+    prev_theta = np.array(theta_0, copy=True)
```
We only really need a single `theta` variable which gets updated in each round. I suggest dropping `theta_0` and `prev_theta`

BGluth replied:
Done.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    theta = None
+    theta_0 = _init_theta_from_moment_accountant(prms.num_features, prms.num_labels)
+    prev_theta = np.array(theta_0, copy=True)
+    user_sel_prob = 1.0 / prms.num_users
```
The selection probability should be a param read from `prms`. It determines the proportion of clients to include in each round, anlagous to `client_fraction` in `simulation_util.server_update`.

BGluth replied:
Changed.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    user_sel_prob = 1.0 / prms.num_users
+    standard_dev = _calc_standard_dev(
+        prms.noise_scale, prms.sensitivity, user_sel_prob, weight_sum
+    )  # This feels weird being a constant...
```
It's constant because it's related to the maximum theoretical contribution of any single client, which is determined by S (`prms.sensitivity`), as opposed to observed contributions.

BGluth replied:
:+1:

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+        for user_idx in random_user_idxs_sample:
+            user_round_labels, user_round_feats = _get_data_for_user_for_round(
+                prms, data, user_idx
+            )
```
This seems to restrict the user's data to a subset. This is not necessary.

BGluth replied:
I'm a bit confused on this one. Like we're just pulling random entries from a user/round right?

Are you thinking about ignoring the grouping of data/user and instead just pulling data randomly from the entire set?

BGluth replied:
Ok I changed it so all of the user's data is used when they are randomly picked.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+            )
+            user_updates_buf.append(
+                user_update_fed_avg(prms, user_round_feats, user_round_labels, theta_0)
+            )
```
This should use `theta`, the current weights, not `theta_0`.

BGluth replied:
Done.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+        )
+
+        theta = prev_theta + merged_user_values + rand_gauss_noise
+        prev_theta = theta
```
As above, don't need to maintain `prev_theta` as a separate variable.

BGluth replied:
Changed.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+
+
+def flat_clip(sensitivity, vec):
+    vec *= min(1, sensitivity / np.linalg.norm(vec))
```
It would be better to use a conditional form, to avoid dividing if the norm is small. Ie:
```python
norm = np.linalg.norm(vec)
if norm > sensitivity:
  vec *= sensitivity / norm
```

BGluth replied:
Optimized!

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    round_num_entries = len(round_user_features)
+
+    # For now just skip any batches that are smaller than batch_size
+    num_batches = round_num_entries // prms.batch_size
```
I don't think there's any reason to drop the last batch, even if it's smaller than the target size.

BGluth replied:
Agreed. Fixed.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+def _get_random_selection_of_user_idxs(num_users):
+    # For now just randomly pick 1/5th of the users.
+    # TODO: Change this later!
+    return np.random.choice(num_users, num_users // 5)
```
This should use `user_sel_prob`.

BGluth replied:
This is now redone. I think you may want to give it a second glance however.

dzeber replied:
I think your updated version is fine. We could optimize this using numpy vectorization but the logic will be the same.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+
+
+def _init_theta_from_moment_accountant(num_features, num_labels):
+    # Just a placeholder value for now
```
As mentioned above, drop the "moment accountant" reference from the function name.

BGluth replied:
Done.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    return np.zeros(num_labels * num_features + num_labels, dtype=np.float64, order="C")
+
+
+def _init_user_weights_and_weight_sum(num_users, weight_mod):
```
In the paper, n_k is the number of training examples for user k. Ie., user's weight is the ratio of the size of their personal dataset to the cap (`weight_mod`), bounded at 1.

I would restrict this function to computing the weight for a single user, given their dataset, and move the rest to the main runner function above. The runner function would then loop over users computing individual user weights and the sum.

BGluth replied:
Ok I made the fixes for this.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    return (noise_scale * sensitivity) / (usr_sel_prob * weight_sum)
+
+
+def _get_data_for_user_for_round(prms, data, user_id):
```
It looks like this limits the data used for a given user in the current round to a random subset of size at least as large as the batch size. This is not necessary - each round should cover the full user data, split into batches (as is done in `user_update_fed_avg`).

BGluth replied:
Done.

dzeber commented on mozfldp/fed_avg_w_dp.py:
```diff
@@ -0,0 +1,250 @@
+    return np.random.normal(loc=0.0, scale=stndrd_dev, size=vec_len)
+
+
+def _choose_n_labels_and_feautures_from_user_labels_and_data(
```
Cf comment for `_get_data_for_user_for_round`: this function shouldn't be necessary.

BGluth replied:
Removed.

dzeber reviewed (changes requested):
Good work! This was a challenging implementation covering a whole bunch of unfamiliar concepts. There are a few points that need to be ironed out, commented inline.

A couple of other general points:
- Make sure to document the code. There are some comments scattered through, but each function should have a docstring, at the very least a one-liner summarizing what it does.
- Rather than using a class just to wrap the parameters, it would be better to have the majority of these functions as methods in there as well, so that they get internal access to the params, rather than still having to pass the param object around. However, let's keep that as a separate issue so as not to block this PR. Also it will depend on the ongoing architecture change to an actual client-server model.
- I don't see a benefit to keep converting between the (coef, intercept) and theta forms for the model weights. Since the `SGDClassifier` uses (coef, intercept), let's just use this throughout. However, let's also break this off into a separate issue after the PR is merged.

BGluth committed:
Now randomly selects users for each round

- Previously always randomly picked 20% of the users.
- Also changed the way user_sel_prob is calculated.

BGluth committed:
Fixed merging user thetas exploding if no user thetas are selected

BGluth committed:
Cleaned up a function

BGluth committed:
Removed prms being passed into local function

- Makes working with the function easier for tests.

BGluth committed:
Renamed "weight_mod" --> "user_weight_cap"

BGluth committed:
Renamed "_init_theta_from_moment_accountant" --> "_init_theta"

- Initializing theta has no releation to the moment accountant.
- Was caused by speed reading two variables being initialized on the same line from the paper. :)

BGluth committed:
`user_sel_prob` is now passed in from the runner

BGluth committed:
Removed comment about standard dev weirdness

BGluth committed:
Removed theta_0 and prev theta

- theta_0 still exists as a name in `user_update_fed_avg` since this is the naming convention in the paper.

BGluth committed:
All user data is now used when a user selected

- Previously pick a random set of data from the user's data.

BGluth committed:
Flat clipping now only does vector mult when needed

- Small performance optimization.

BGluth committed:
Now allow batches that are smaller than batch size

BGluth committed:
Corrected the way user weights are calculated

BGluth commented:
Ok @dzeber. I made all the requested changes. Ball is back in your court.

BGluth committed:
Added comments... Probably too many...

- All functions (except TODO ones) now have doc strings.

Commit:
Update dataset path in runner notebook

dzeber commented:
Your fixes look good. I think this is ready to go.
